#summary How to install the MetaProteomeAnalyzer on a linux server.
#labels Featured
#sidebar TableOfContents

<table>
  <tr>
    <td width="70%">
 * [#General_Information General Information]
 * [#System_Requirements System Requirements]
 * [#SQL_Database_Setup SQL Database Setup]
 * [#Algorithms_Installation Algorithms Installation]
 * [#FASTA_Database_Formatting FASTA Database Formatting]
 * [#Server_Startup Server Startup]
    </td>
  </tr>
</table>


----
==General Information==

On this page you will find information how to setup the MetaProteomeAnalyzer server software, which is available via the `mpa-server-X.Y.Z.zip` file (see Downloads section).

*Important:* Please note that you may require some IT administration staff (and rights) in order to successfully install the MetaProteomeAnalyzer server pipeline.

[#General_Information Go to top of page]
----
==System Requirements==

 * Linux/Ubuntu system (preferably)
 * MySQL database 
 * Minimum 2 GB RAM
 * Fast 64-bit machine (though 32-bit is working as well)

*System folders:*

The default folder structure we used for our *metaprot* server was the following:
{{{
/scratch/metaprot/software (The main folder for the bash scripts and the mpa-server-X.Y.Z.jar)
}}}

For the data exchange (received spectra from the client and identification result files) we recommend to also use a default like this:
{{{
/scratch/metaprot/data
}}}

However, it is possible to adjust these folders by the provided server properties file.

[#General_Information Go to top of page]
----
==SQL Database Setup==
Make sure you have a ready MySQL server running and create the table *metaprot*.
Then execute the provided SQL structure script called `metaprot.sql` from the folder *SQL database* (in the unzipped mpa-server folder structure).

After execution 26 tables should have been created.

The next step is to insert data into the taxonomy table by using the script `inserttaxa.sh` from the *scripts* folder.
*Important:* Please make sure that the table *taxonomy* in the SQL metaprot database has been successfully filled.

Finally, you need to adjust your MySQL user name and password in the `DbConnectionSettings.txt` file and move this file to the 
default password folder:
{{{
/scratch/metaprot/software/password 
}}}
*Important:* The password folder should be located in the software folder where mpa-server-X.Y.Z.jar is being executed.

Note: If you have any trouble with setting up the MySQL database, please contact your local system administrator.

[#General_Information Go to top of page]
----
==Algorithms Installation==

Please note that the installation instructions may not be up-to-date. Please visit the websites of the search engine algorithms to get the latest versions and further installation details.

 * X!Tandem [http://www.thegpm.org/tandem/ Link to website]
{{{

Installation:

wget ftp://ftp.thegpm.org/projects/tandem/source/tandem-linux-13-09-01-1.tar.gz
tar xvfz tandem-linux-13-09-01-1.tar.gz

Note: There are 32-bit and 64-bit versions of X!Tandem available. Please choose the version which is suitable for your system.

Folder:
/scratch/metaprot/software/xtandem

Example usage:
./xtandem.exe input.xml
}}}

 * OMSSA [ftp://ftp.ncbi.nih.gov/pub/lewisg/omssa/CURRENT/ Link FTP Server]
{{{
Installation:
wget ftp://ftp.ncbi.nih.gov/pub/lewisg/omssa/CURRENT/omssa-linux.tar.gz
tar xvfz omssa-linux.tar.gz

Folder:
/scratch/metaprot/software/omssa

Note: You need to install a version of BLAST for formatting of the FASTA databases.

Installation BLAST:
_32-bit:_
wget ftp://ftp.ncbi.nih.gov/blast/executables/LATEST/ncbi-blast-2.2.28+-ia32-linux.tar.gz 
tar xvfz ncbi-blast-2.2.28+-ia32-linux.tar.gz

_64-bit:_
wget ftp://ftp.ncbi.nih.gov/blast/executables/LATEST/ncbi-blast-2.2.28+-x64-linux.tar.gz 
tar xvfz ncbi-blast-2.2.28+-x64-linux.tar.gz

Folder:
/scratch/metaprot/software/blast
}}}

 * Crux [http://noble.gs.washington.edu/proj/crux/ Link to website]
{{{
Installation:
_32-bit:_
wget http://noble.gs.washington.edu/proj/crux/download/crux-1.40.Linux.i686.zip
unzip crux-1.40.Linux.i686.zip

_64-bit:_
wget http://noble.gs.washington.edu/proj/crux/download/crux-1.40.Linux.x86_64.zip
unzip crux-1.40.Linux.x86_64.zip

Folder:
/scratch/metaprot/software/crux
}}}
*Important:* Please make sure you use a parameters file for Crux and put it into the Crux bin-folder here:
/scratch/metaprot/software/crux/bin

An example of a default parameters file can be found [http://noble.gs.washington.edu/proj/crux/default.params here (default.params)].


 * Inspect [http://proteomics.ucsd.edu/Software/Inspect.html Link to website]
{{{
Installation:
wget http://proteomics.ucsd.edu/Downloads/Inspect.20120109.zip
mkdir /scratch/metaprot/software/inspect (Attention: the zipfile does not contain a separate folder)
Move to this folder and unzip the file to it:
unzip Inspect.20120109.zip

Dependencies: 
_expat_
sudo apt-get install expat libexpat1-dev (Ubuntu)
_python-numpy_
apt-get install python-numpy (Ubuntu)

Folder:
/scratch/metaprot/software/inspect
}}}
 
[#General_Information Go to top of page]
----
==FASTA Database Formatting==
The FASTA databases (e.g. uniprot_sprot.fasta) should be provided to the default FASTA database folder:
{{{
/scratch/metaprot/data/fasta
}}}

*Important:*
If you have your FASTA file(s) in place, copy the following files and directories from the folder *FASTA* (from the mpa-server zipfile) to the  FASTA database folder (see above);
 * `indexfasta.jar`
 * `fastaformat.sh`
 * `DBTools` (+ subdirectories)

The actual FASTA formatting in then executing by the following command
{{{
./fastaformat.sh uniprot_sprot 
Important: Please note that the file ending *.fasta must be omitted here!
}}}

Note: Please edit the script to your demands, e.g. by omitting certain search engine algorithms.

[#General_Information Go to top of page]
----
==Server Startup==
Move the scripts `clearfolders.sh` and `startserver.sh` from the folder *scripts* (from the mpa-server zipfile) to the default software folder (e.g. /scratch/metaprot/software).

Finally, in this folder the server can be started via a startup script `startserver.sh` or the following command:
{{{
nohup java -jar mpa-server-X.Y.Z.jar -XmxXXXXm where XXXX stands for maximum assigned memory (in megabytes of RAM)
}}}

[#General_Information Go to top of page]
----